In [16]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score



In [96]:
data = pd.read_csv(r"C:\Users\KRP\Programming\School\AiCphBusiness\MachineLearning\Assignment2\diamonds.csv")
data = pd.get_dummies(data, columns= ["color","clarity", "cut"])
y = data["price"]
x = data.drop(["price", "Unnamed: 0"], axis =1)
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3)


To Test the for over fitting, we can use cross validation. Overfitting is when the model is trained too well on the training dataset that it is not able predict datapoints outside of the training set. Cross validation includes resampling the training data set and splitting it into different portions. The model is then trained on the training set several times, excluding the validation sets. The model is then tested on the different cross validation sets to see if it statistically performs differently. 

In [94]:

lm = LinearRegression()
# Perform 5-fold cross-validation
cv_scores = cross_val_score(lm, x_train, y_train, cv=5,scoring= "r2")
# Print cross-validation scores
print(f"CV Scores: {cv_scores}")
# Average CV score
print(f"Average CV Score: {np.mean(cv_scores)}")

CV Scores: [0.91858432 0.92105606 0.91808091 0.92015715 0.91717296]
Average CV Score: 0.9190102806107013


In [95]:
lm = LinearRegression()
scores = cross_val_score(lm, x, y, cv=5, scoring='r2')

# Print cross-validation scores
print(f"CV Scores: {scores}")
# Average CV score
print(f"Average CV Score: {np.mean(scores)}")

CV Scores: [  0.34224026   0.70010775   0.83186637 -14.25049522  -0.86365774]
Average CV Score: -2.647987716317788


In [22]:
y_pred = model.predict(x_test)

print(y_pred)

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = mean_squared_error(y_test, y_pred, squared=False) 
r2 = r2_score(y_test, y_pred)

print(f"MAE: {mae}")
print(f"MSE: {mse}")
print(f"RMSE: {rmse}")
print(f"R-squared: {r2}")

[  711.56570501  3197.40419579  1943.02948182 ... 10609.79287119
  4105.70876126  1993.88821027]
MAE: 734.2415163929631
MSE: 1244765.4357158723
RMSE: 1115.6905645006918
R-squared: 0.9201866914388085


C:\Users\KRP\Programming\School\AiCphBusiness\.venv\Lib\site-packages\sklearn\metrics\_regression.py:483: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


MAE tells you, on average, how much the predictions deviate from the actual values. An MAE of 734.24 suggests that, on average, the model's predictions are about 734 units away from the actual values. The interpretation of this number depends on the context of your problem and the scale of the target variable.

MSE is the average of the squares of the errors. It's more sensitive to outliers than MAE because it squares the differences. Your MSE value is relatively large because it's influenced by the scale of your data and the squared differences.

RMSE, which is the square root of MSE, is often more interpretable than MSE because it's in the same units as your target variable. An RMSE of 1,115.69 indicates the standard deviation of the prediction errors. Like the MAE, whether this is considered high or low depends on the scale of your target variable and the context of your problem.

R-squared value of 0.92 is quite high, indicating that your model explains 92% of the variance in the target variable based on its inputs. This suggests that the model fits the data well and is capable of capturing most of the variability in the target variable with its predictions.

In [25]:
y_pred = model.predict(x_train)

print(y_pred)

mae = mean_absolute_error(y_train, y_pred)
mse = mean_squared_error(y_train, y_pred)
rmse = mean_squared_error(y_train, y_pred, squared=False)
r2 = r2_score(y_train, y_pred)

print(f"MAE: {mae}")
print(f"MSE: {mse}")
print(f"RMSE: {rmse}")
print(f"R-squared: {r2}")

[8059.33067384  606.62346058 8567.48039848 ... 1739.45892698 2501.07703433
 6787.8108665 ]
MAE: 746.033190874007
MSE: 1290541.9147450614
RMSE: 1136.0202087749415
R-squared: 0.9195976267987521


C:\Users\KRP\Programming\School\AiCphBusiness\.venv\Lib\site-packages\sklearn\metrics\_regression.py:483: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(
